common4j/src/main/com/microsoft/identity/common/java/net/UrlConnectionHttpClient.java

```diff
@@ -0,0 +1,361 @@
+                // timeout, we want to retry once. Throw the exception to the upper layer, and the
+                // upper layer will handle the retry.
+                throw socketTimeoutException;
+            } catch (final IOException ioException) {
+                // 404, for example, will generate an exception.  We should catch it.
+                responseStream = urlConnection.getErrorStream();
+            }
+
+            final int statusCode = urlConnection.getResponseCode();
+            final Date date = new Date(urlConnection.getDate());
+
+            final String responseBody = responseStream == null
+                    ? ""
+                    : convertStreamToString(responseStream);
+
+            response = new HttpResponse(
+                    date,
+                    statusCode,
+                    responseBody,
+                    urlConnection.getHeaderFields()
+            );
+        } finally {
+            completionCallback.accept(response);
+            safeCloseStream(responseStream);
+        }
+
+        return response;
+    }
+
+    private HttpURLConnection setupConnection(HttpRequest request) throws IOException {
+        final HttpURLConnection urlConnection = HttpUrlConnectionFactory.createHttpURLConnection(request.getRequestUrl());
+
+        // Apply request headers and update the headers with default attributes first
+        final Set<Map.Entry<String, String>> headerEntries = request.getRequestHeaders().entrySet();
+
+        for (final Map.Entry<String, String> entry : headerEntries) {
+            urlConnection.setRequestProperty(entry.getKey(), entry.getValue());
+        }
+
+        urlConnection.setConnectTimeout(getConnectTimeoutMs());
+        urlConnection.setReadTimeout(getReadTimeoutMs());
+        urlConnection.setInstanceFollowRedirects(true);
+        urlConnection.setUseCaches(false);
+        urlConnection.setDoInput(true);
+
+        return urlConnection;
+    }
+
+    private Integer getReadTimeoutMs() {
+        return readTimeoutMsSupplier == null ? readTimeoutMs : readTimeoutMsSupplier.get();
+    }
+
+    private Integer getConnectTimeoutMs() {
+        return connectTimeoutMsSupplier == null ? connectTimeoutMs : connectTimeoutMsSupplier.get();
+    }
+
+    private static void setRequestBody(@NonNull final HttpURLConnection connection,
+                                       final byte[] contentRequest,
+                                       final String requestContentType) throws IOException {
+        if (contentRequest == null) {
+            return;
+        }
+
+        connection.setDoOutput(true);
+
+        if (!StringUtil.isNullOrEmpty(requestContentType)) {
+            connection.setRequestProperty("Content-Type", requestContentType);
+        }
+
+        connection.setRequestProperty("Content-Length", String.valueOf(contentRequest.length));
+
+        OutputStream out = null;
+
+        try {
+            out = connection.getOutputStream();
```

Comment:
>out = connection.getOutputStream(); [](http://example.com/codeflow?start=12&length=35)

if this fails or throws.. does it set the response/input stream accordingly?
Do we have test cases covering that?

Reply:
was just looking at this.. [https://stackoverflow.com/questions/4816824/httpurlconnection-post-conn-getoutputstream-throwing-exception](https://stackoverflow.com/questions/4816824/httpurlconnection-post-conn-getoutputstream-throwing-exception)

Reply:
We're currently throwing IOException, which is base class of the SocketException.

Those exceptions are not handled in this layer.